streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessor.java

```diff
@@ -17,70 +17,158 @@
+                        }
+                        return;
+                    case SHUT_DOWN:
+                        throw new StreamsException(String.format(
+                            "%s buffer exceeded its max capacity [%d/%d] records and [%d/%d] bytes",
+                            internalProcessorContext.currentNode().name(),
+                            buffer.numRecords(), maxRecords,
+                            buffer.bufferSize(), maxBytes
+                        ));
+                }
+            }
+        }
     }
 
-    @Override
-    public String toString() {
-        return "KTableSuppressProcessor{" +
-            "suppress=" + suppress +
-            ", keySerde=" + keySerde +
-            ", valueSerde=" + valueSerde +
-            '}';
+    private void drainExpiredRecords(
+        final long expiryTime,
+        final Iterator<KeyValue<TimeKey<Bytes>, ContextualRecord<byte[]>>> iterator) {
+
+        while (iterator.hasNext() && buffer.minTimestamp() <= expiryTime) {
+            final KeyValue<TimeKey<Bytes>, ContextualRecord<byte[]>> next = iterator.next();
+            evictAndEmit(iterator, next);
+        }
+    }
+
+    private void evictAndEmit(final Iterator<KeyValue<TimeKey<Bytes>, ContextualRecord<byte[]>>> iterator,
+                              final KeyValue<TimeKey<Bytes>, ContextualRecord<byte[]>> next) {
+        final Bytes key = next.key.key();
+        final ContextualRecord<Change<V>> contextualRecord = new ContextualRecord<>(
+            valueSerde.deserializer().deserialize(null, next.value.value()),
+            next.value.recordContext()
+        );
+        // to be safe, forward before we delete
+        setNodeAndForward(key, contextualRecord);
+        iterator.remove();
     }
 
-    public static class NotImplementedException extends RuntimeException {
-        NotImplementedException() {
-            super();
+    private boolean overCapacity() {
+        return buffer.numRecords() > maxRecords || buffer.bufferSize() > maxBytes;
+    }
+
+    private void setNodeAndForward(final Bytes key, final ContextualRecord<Change<V>> contextualRecord) {
+        if (shouldForward(contextualRecord.value())) {
+            final ProcessorRecordContext prevRecordContext = internalProcessorContext.recordContext();
+            internalProcessorContext.setRecordContext(contextualRecord.recordContext());
+            try {
+                final K deserializedKey = keySerde.deserializer().deserialize(null, key.get());
+                internalProcessorContext.forward(deserializedKey, contextualRecord.value());
+            } finally {
+                internalProcessorContext.setRecordContext(prevRecordContext);
+            }
         }
     }
+
+    private boolean shouldForward(final Change<V> value) {
+        return !(value.newValue == null && shouldSuppressTombstones);
+    }
+
+    @Override
+    public void close() {
+    }
```

Comment:
Should we clear the buffer upon closing? Maybe it does not make a difference on correctness, but would it worthy for performance?

Reply:
Yeah, I wasn't sure. I'll go ahead and do it.

Reply:
Actually, let's defer this to Part 4, where the buffer becomes a proper store, and has its own `close()` method.